wait_until_replica_is_caught_up_to_master.go

```diff
@@ -0,0 +1,125 @@
+package ghostferry
+
+import (
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/siddontang/go-mysql/mysql"
+	"github.com/sirupsen/logrus"
+)
+
+type ReplicatedMasterPositionFetcher interface {
+	Current(*sql.DB) (mysql.Position, error)
+}
+
+// Selects the master position that we have replicated until from a heartbeat
+// table of sort.
+type ReplicatedMasterPositionViaCustomQuery struct {
+	// The custom query executing should return a single row with two values:
+	// the string file and the integer position. For pt-heartbeat, this query
+	// would be:
+	//
+	// "SELECT file, position FROM meta.ptheartbeat WHERE server_id = %d" % serverId
+	//
+	// where serverId is the master server id, and meta.ptheartbeat is the table
+	// where pt-heartbeat writes to.
+	//
+	// For pt-heartbeat in particular, you should not use the
+	// relay_master_log_file and exec_master_log_pos of the DB being replicated
+	// as these values are not the master binlog positions.
+	Query string
+}
+
+func (r ReplicatedMasterPositionViaCustomQuery) Current(replicaDB *sql.DB) (mysql.Position, error) {
+	var file string
+	var pos uint32
+	row := replicaDB.QueryRow(r.Query)
+	err := row.Scan(&file, &pos)
+	if err != nil {
+		return mysql.Position{}, nil
+	}
+
+	switch {
+	case err == sql.ErrNoRows:
+		return mysql.Position{}, fmt.Errorf("no results from query")
+	case err != nil:
+		return mysql.Position{}, err
+	default:
+		if file == "" {
+			return mysql.Position{}, fmt.Errorf("query does not return a file")
+		}
+
+		return mysql.Position{
+			Name: file,
+			Pos:  pos,
+		}, nil
+	}
+}
+
+// Only set the MasterDB and ReplicatedMasterPosition options in your code as
+// the others will be overwritten by the ferry.
+type WaitUntilReplicaIsCaughtUpToMaster struct {
```

Comment:
maybe name this `SlaveReplicationWaiter` in the spirit of the other vocabulary suggestion (`SlavePosition`)?